SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialVirtualWorkerTranslationManager.cpp

```diff
@@ -0,0 +1,190 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "EngineClasses/SpatialVirtualWorkerTranslationManager.h"
+#include "EngineClasses/SpatialVirtualWorkerTranslator.h"
+#include "Interop/Connection/SpatialWorkerConnection.h"
+#include "Interop/SpatialReceiver.h"
+#include "SpatialConstants.h"
+#include "Utils/SchemaUtils.h"
+
+DEFINE_LOG_CATEGORY(LogSpatialVirtualWorkerTranslationManager);
+
+SpatialVirtualWorkerTranslationManager::SpatialVirtualWorkerTranslationManager(
+	USpatialReceiver* InReceiver,
+	USpatialWorkerConnection* InConnection,
+	SpatialVirtualWorkerTranslator* InTranslator)
+	: Receiver(InReceiver)
+	, Connection(InConnection)
+	, Translator(InTranslator)
+	, bWorkerEntityQueryInFlight(false)
+{}
+
+void SpatialVirtualWorkerTranslationManager::AddVirtualWorkerIds(const TSet<VirtualWorkerId>& InVirtualWorkerIds)
+{
+	// Currently, this should only be called once on startup. In the future we may allow for more
+	// flexibility. 
+	check(UnassignedVirtualWorkers.IsEmpty());
+	for (VirtualWorkerId VirtualWorkerId : InVirtualWorkerIds)
+	{
+		UnassignedVirtualWorkers.Enqueue(VirtualWorkerId);
+	}
+}
+
+void SpatialVirtualWorkerTranslationManager::AuthorityChanged(const Worker_AuthorityChangeOp& AuthOp)
+{
+	const bool bAuthoritative = AuthOp.authority == WORKER_AUTHORITY_AUTHORITATIVE;
+
+	if (!bAuthoritative)
+	{
+		UE_LOG(LogSpatialVirtualWorkerTranslationManager, Error, TEXT("Lost authority over the translation mapping. This is not supported."));
+		return;
+	}
+
+	check(AuthOp.component_id == SpatialConstants::VIRTUAL_WORKER_TRANSLATION_COMPONENT_ID);
+	UE_LOG(LogSpatialVirtualWorkerTranslationManager, Log, TEXT("This worker now has authority over the VirtualWorker translation."));
+
+	// TODO(zoning): The prototype had an unassigned workers list. Need to follow up with Tim/Chris about whether
+	// that is necessary or we can continue to use the (possibly) stale list until we receive the query response.
+
+	// Query for all connection entities, so we can detect if some worker has died and needs to be updated in
+	// the mapping.
+	QueryForWorkerEntities();
+}
+
+// For each entry in the map, write a VirtualWorkerMapping type object to the Schema object.
+void SpatialVirtualWorkerTranslationManager::WriteMappingToSchema(Schema_Object* Object) const
+{
+	for (auto& Entry : VirtualToPhysicalWorkerMapping)
```

Comment:
Feel like this could be const& Not sure on schema api

Reply:
ok